x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/HashAggregationOperator.java

```diff
@@ -156,16 +167,21 @@ public void close() {
                     Releasables.closeExpectNoException(prepared);
                 }
             }
+            Block[] keys = new Block[groups.size()];
+            page = wrapPage(page);
             try (AddInput add = new AddInput()) {
-                checkState(needsInput(), "Operator is already finishing");
-                requireNonNull(page, "page is null");
+                for (int g = 0; g < groups.size(); g++) {
+                    keys[g] = groups.get(g).eval(page);
+                }
 
                 for (int i = 0; i < prepared.length; i++) {
                     prepared[i] = aggregators.get(i).prepareProcessPage(blockHash, page);
                 }
 
-                blockHash.add(wrapPage(page), add);
+                blockHash.add(new Page(keys), add);
```

Comment:
I'd like to modify `BlockHash` to take a `Block[]` with the blocks in the right position. But that seems like something for another time.